### PR TITLE
yarnInstallHook: init

### PR DESCRIPTION
**Pull request by lelgenio**
## Description of changes

When attempt to use `npmInstallHook` on a yarn project using `fetchYarnDeps`, npm frequently tries to connect to the internet during the `prune` step.

Turns out replacing `npm prune` with `yarn install --production=true ...` resolves this issue. 

ping @doronbehar

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 328544` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>blade-formatter</li>
    <li>codefresh</li>
    <li>element-call</li>
    <li>postlight-parser</li>
    <li>yarnInstallHook</li>
  </ul>
</details>

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:47]: Could you explain these conditionals?

1. Wouldn't the commands below fail more informatively if the node_modules directory wouldn't exist and you wouldn't silently don't do anything in that case?
2. `dontYarnPrune` is not documented, and after reading the comment it seems confusing to me - we have `dontYarnInstall` already which achieves the same behavior.

**Review by doronbehar — COMMENTED** [review]
Looks good, thanks for working on this!

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:47]: > Could you explain these conditionals?
> 
>     1. Wouldn't the commands below fail more informatively if the node_modules directory wouldn't exist and you wouldn't silently don't do anything in that case?

This is a vestige from npmInstallHook, `npm prune` fails if `node_modules` does not exist AFIK. 
Still, I think it makes sense to keep this as is, `node_modules` should be created by `yarnConfigHook`, in the edge case that `yarnInstallHook` does not see one, I think it would be unexpected for it to create a "new" `node_modules` by itself. 
If there was a `yarn prune` command, I would be using that, and then, this check would be mandatory, is what I'm trying to say.

>     2. `dontYarnPrune` is not documented, and after reading the comment it seems confusing to me - we have `dontYarnInstall` already which achieves the same behavior.

The documentation is [here](https://github.com/NixOS/nixpkgs/pull/328544/files#diff-69c4ba235ab61f76ad80555351e0567178235fce8b8f966ff689e758e1beb771R445), I added it later.

We need both `dontYarnInstall` and `dontYarnPrune` to support packages that incorrectly put runtime dependencies in `devDependencies`. Ideally we wouldn't need `dontYarnPrune`, but I'm not sure we can remove it, so I'm playing it safe.

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:46]: `dontYarnInstall`?

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:80]: Or `dontYarnPrune`?

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [inline doc/languages-frameworks/javascript.section.md:384]: Nit: This commit could be squashed to the 1st commit.

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [inline doc/languages-frameworks/javascript.section.md:445]: I am still confused regarding this... https://yarnpkg.com/cli/install doesn't mention the word `devDependencies`, or the word prune. It'd also be nice to find a single example where this argument is needed, and explain like that why is that.

**Comment by winterqt** [comment]
I have no spoons for this, sorry.

File: doc/languages-frameworks/javascript.section.md
Comment (lelgenio) [reply doc/languages-frameworks/javascript.section.md:445]: Here ya go: 
https://classic.yarnpkg.com/en/docs/cli/prune
https://classic.yarnpkg.com/en/docs/cli/install#toc-yarn-install-production-true-false

> It'd also be nice to find a single example where this argument is needed, and explain like that why is that.

I'm not sure what you mean here. Is it:

A) "dontYarnPrune doesn't seem very useful, we should always prune"? 

I left the option to disable pruning because I don't trust that yarn will not suffer from the same problem as already exists in npm — not being able to prune without internet access sometimes.

Like other cleanup steps (stripping binaries, for example), this feature is nice to have, but imprudent to not allow others to easily disable it.

B) "prune doen't seem very useful, we should not do it"

```sh
$ du -sh result-not-pruned/ result-pruned/
702M	result-not-pruned/
120M	result-pruned/
```

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:80]: Both, they do different things, and there's two flags on purpose.

Looking at the code might make things more confusing since `dontYarnPrune` skips running `yarn install` and `dontYarnInstall` skips running things which are not `yarn install`.

It might be better to think of the abstraction that's being created here:

- `yarnInstallHook` cleans up devDependencies, installs executables and man pages.
- Don't want to run any of that? `dontYarnInstall = true`
- Don't want to remove devDependencies, but otherwise run the normal installation? `dontYarnPrune = true`

File: doc/languages-frameworks/javascript.section.md
Comment (doronbehar) [reply doc/languages-frameworks/javascript.section.md:445]: What confuses me is only the terminology you use for the different `dontYarn*` variables, and not the logic you implement. 

> > It'd also be nice to find a single example where this argument is needed, and explain like that why is that.
> 
> I'm not sure what you mean here. Is it:

Indeed my phrasing wasn't clear. What I meant was: Give me 1 example for a package that needs to set `dontYarnPrune = true`, and show me how the `yarn install` command fails - with which errors, and maybe I'll get convinced that it'd be more correct to disable that `yarn install` command with a variable named `dontYarnPrune`, and not `dontYarnInstall`. Both of your A & B arguments you gave are very convincing, but still don't convince me that the naming is good.

> Here ya go:
> [classic.yarnpkg.com/en/docs/cli/prune](https://classic.yarnpkg.com/en/docs/cli/prune)
> [classic.yarnpkg.com/en/docs/cli/install#toc-yarn-install-production-true-false](https://classic.yarnpkg.com/en/docs/cli/install#toc-yarn-install-production-true-false)

Again, your shell code links the shell variable `dontYarnPrune` to the command `yarn install`. Hence I seek to find a connection / relation between the functionality of the `yarn install` command, and the word "prune". The 1st link is not the documentation of the command `yarn install` which you use right after the `dontYarnPrune` check, and the 2nd link doesn't mention the word prune.

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:80]: > Both, they do different things, and there's two flags on purpose.
> 
> Looking at the code might make things more confusing since `dontYarnPrune` skips running `yarn install` and `dontYarnInstall` skips running things which are not `yarn install`.

I'd suggest a different naming:

- Don't check `dontYarnInstall` here - where you set `installPhase`.
- Check `dontYarnInstall` only in the `yarnInstallHook` - where now you check `dontInstallPrune`.

If people will want from some reason to add the `yarnInstallHook` to their `nativeBuildInputs` and not use the hook directly (or not use it at all), they can always override the `installPhase`.

Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:80]: I don't like this, I've personally never seen a hook that cannot be disabled like this.
Also other yarn hooks follow the same nomenclature dontYarn{Configure,Build}, I don't see how any name other than `dontYarnInstall` would make sense.

The name `dontYarnPrune` was chosen to follow `dontNpmPrune`.

I think we're talking about the same thing on two threads :smiley:

File: doc/languages-frameworks/javascript.section.md
Comment (lelgenio) [reply doc/languages-frameworks/javascript.section.md:445]: > Again, your shell code links the shell variable `dontYarnPrune` to the command `yarn install`. Hence I seek to find a connection / relation between the functionality of the `yarn install` command, and the word "prune". The 1st link is not the documentation of the command `yarn install` which you use right after the `dontYarnPrune` check, and the 2nd link doesn't mention the word prune.

Npm has a command called `prune`: https://docs.npmjs.com/cli/v10/commands/npm-prune

In Yarn, the functionality of this command is bundled in the `install` command. You can see this in the first link "The prune command isn’t necessary. yarn install will prune extraneous packages.", meaning the functionality of a `prune` command is bundled in the `install` command.

I think my usage of `yarn install --production=true` is the closes possible to what `yarn prune` would do, if it was a separate command (right now it just returns an error message saying you should use `yarn install`).

Comment (doronbehar) [reply doc/languages-frameworks/javascript.section.md:445]: > I think we're talking about the same thing on two threads 😃

Correct :).

> Npm has a command called `prune`: [docs.npmjs.com/cli/v10/commands/npm-prune](https://docs.npmjs.com/cli/v10/commands/npm-prune)
> 
> In Yarn, the functionality of this command is bundled in the `install` command. You can see this in the first link "The prune command isn’t necessary. yarn install will prune extraneous packages.", meaning the functionality of a `prune` command is bundled in the `install` command.

OK I see that now. I missed that text when I opened the link last time.

> I think my usage of `yarn install --production=true` is the closes possible to what `yarn prune` would do,

Hmm OK I see. After testing building `postlight-master` with and without `dontYarnPrune` set and I see a difference that can be explained by the `--production=true` flag, and I _am_ convinced that we should run `yarn install --production=true` to reduce the amount of files in the `node_modules` directory.

What if we'd rename the `dontYarnPrune` variable to `yarnIncludeDevDeps`? This way you'd be more aligned with upstream's terminology.

Comment (lelgenio) [reply doc/languages-frameworks/javascript.section.md:445]: > What if we'd rename the `dontYarnPrune` variable to `yarnIncludeDevDeps`? This way you'd be more aligned with upstream's terminology.

That sound alright, maybe even even `yarnKeepDevDeps`, since they will always be installed, and then optionally removed?

Comment (doronbehar) [reply doc/languages-frameworks/javascript.section.md:445]: > > What if we'd rename the `dontYarnPrune` variable to `yarnIncludeDevDeps`? This way you'd be more aligned with upstream's terminology.
> 
> That sound alright, maybe even even `yarnKeepDevDeps`, since they will always be installed, and then optionally removed?

Sounds even better.

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 328544` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>blade-formatter</li>
    <li>codefresh</li>
    <li>nixpkgs-manual</li>
    <li>postlight-parser</li>
    <li>yarnInstallHook</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
> When attempt to use `npmInstallHook` on a yarn project using `fetchYarnDeps`, npm frequently tries to connect to the internet during the `prune` step.
> 
> Turns out replacing `npm prune` with `yarn install --production=true ...` resolves this issue.
> 
> Right now this is just a copy-paste of `npmInstallHook`, but I expect some other tweaks to be necessary, so I did not try to modify `npmInstallHook` to work with yarn.

Only now I noticed the code duplication (sorry for my lack of concentration). I wish we could avoid it, as I doubt we'd ever need to modify yarn packages' manual pages and binaries installation (based on `jq` parsing of `package.json`) differently from other npm packages.

Perhaps we can make this `yarnInstallHook` essentially run `npmInstallHook` with `dontNpmPrune`, and then run `yarn install --production=true`? Also, note how in principal you should mention (somehow) all variables inherited from `npmInstallHook` such as `npmPackFlags`, `npmFlags` which are also used by the new hook you add.

Another option would be to remove the `npm prune` command from `npmInstallHook`, and add an `npmPruneHook`. Then we should do something like this:

```diff
diff --git i/pkgs/build-support/node/build-npm-package/default.nix w/pkgs/build-support/node/build-npm-package/default.nix
index 4bcbf3f14f2f..ca5df6f0f372 100644
--- i/pkgs/build-support/node/build-npm-package/default.nix
+++ w/pkgs/build-support/node/build-npm-package/default.nix
@@ -55,6 +55,8 @@
 , npmBuildHook ? null
   # Custom npmInstallHook
 , npmInstallHook ? null
+  # Custom npmPruneHook
+, npmPruneHook ? null
 , ...
 } @ args:
 
@@ -74,6 +76,7 @@ stdenv.mkDerivation (args // {
       (if npmConfigHook != null then npmConfigHook else npmHooks.npmConfigHook)
       (if npmBuildHook != null then npmBuildHook else npmHooks.npmBuildHook)
       (if npmInstallHook != null then npmInstallHook else npmHooks.npmInstallHook)
+      (if npmPruneHook != null then npmPruneHook else npmHooks.npmPruneHook)
       nodejs.python
     ]
     ++ lib.optionals stdenv.isDarwin [ cctools ];
```

And add `npmPruneHook` manually to expressions that use `npmInstallHook` directly.

What do you think?

**Comment by lelgenio** [comment]
> Only now I noticed the code duplication (sorry for my lack of concentration). I wish we could avoid it, as I doubt we'd ever need to modify yarn packages' manual pages and binaries installation (based on `jq` parsing of `package.json`) differently from other npm packages.
> 
> Perhaps we can make this `yarnInstallHook` essentially run `npmInstallHook` with `dontNpmPrune`, and then run `yarn install --production=true`? Also, note how in principal you should mention (somehow) all variables inherited from `npmInstallHook` such as `npmPackFlags`, `npmFlags` which are also used by the new hook you add.

One thing I still need to figure out is if we maybe need to use `yarn pack` instead of `npm pack` (to be able to pass yarn flags reliably to all yarn* hooks), the installation of executables and manuals will probably stay the same. 

If we are to share code between `npmInstallHook` and `yarnInstallHook`, we should probably not just inherit and instead be explicit in structuring the code in a way that screams "this code is used by more than just npmInstallHook".

> Another option would be to remove the `npm prune` command from `npmInstallHook`, and add an `npmPruneHook`. Then we should do something like this:
> 
> ```diff
> diff --git i/pkgs/build-support/node/build-npm-package/default.nix w/pkgs/build-support/node/build-npm-package/default.nix
> index 4bcbf3f14f2f..ca5df6f0f372 100644
> --- i/pkgs/build-support/node/build-npm-package/default.nix
> +++ w/pkgs/build-support/node/build-npm-package/default.nix
> @@ -55,6 +55,8 @@
>  , npmBuildHook ? null
>    # Custom npmInstallHook
>  , npmInstallHook ? null
> +  # Custom npmPruneHook
> +, npmPruneHook ? null
>  , ...
>  } @ args:
>  
> @@ -74,6 +76,7 @@ stdenv.mkDerivation (args // {
>        (if npmConfigHook != null then npmConfigHook else npmHooks.npmConfigHook)
>        (if npmBuildHook != null then npmBuildHook else npmHooks.npmBuildHook)
>        (if npmInstallHook != null then npmInstallHook else npmHooks.npmInstallHook)
> +      (if npmPruneHook != null then npmPruneHook else npmHooks.npmPruneHook)
>        nodejs.python
>      ]
>      ++ lib.optionals stdenv.isDarwin [ cctools ];
> ```
> 
> And add `npmPruneHook` manually to expressions that use `npmInstallHook` directly.
> 
> What do you think?

You know how sometimes it feels more like you are discovering a solution rather than creating it?
Now that you mention it, it seems clear to me that there's a problem with `npmInstallHook`:
The behavior of pruning should be in a separate hook, probably called `npmFixupHook`, so it matches with a phase. "fixing" this is not vital to this PR though. (A draft: https://github.com/NixOS/nixpkgs/pull/335504)

Maybe we can create a "shared" set of these functions (`installNodeMan`, `wrapNodePrograms` or something to that effect?), use them in `yarnInstallHook`, create a `yarnFixupHook` too.
At first I was skeptical of making changes to `npmInstallHook` in this PR, but if I make it easily revertible (separate commit), it should be fine.

By now each hook would be tiny, and very specific, which is good I think. But there will be too many hooks for it to be reasonable for normal usage, I see a `buildYarnPackage` in the future..

**Comment by doronbehar** [comment]
> One thing I still need to figure out is if we maybe need to use `yarn pack` instead of `npm pack` (to be able to pass yarn flags reliably to all yarn* hooks), the installation of executables and manuals will probably stay the same.

That sounds more correct, and it will also justify separating the hooks.

> Maybe we can create a "shared" set of these functions (`installNodeMan`, `wrapNodePrograms` or something to that effect?), use them in `yarnInstallHook`, create a `yarnFixupHook` too.

After speculating the idea suggested by me and implemented in https://github.com/NixOS/nixpkgs/pull/335504 , I was thinking of another option: Using Nix to share the code! Here's a draft:

- https://github.com/NixOS/nixpkgs/pull/335523

**Comment by doronbehar** [comment]
BTW Another option is to use the plural `postInstallHooks`, as done in the `yarnConfigHook` commit at https://github.com/NixOS/nixpkgs/pull/335751 . See https://github.com/NixOS/nixpkgs/commit/28e5a20afbfeb2cd4c1e4249bebc89cd1b261de3 ( Hopefully this link won't die at some point).

**Comment by lelgenio** [comment]
https://github.com/NixOS/nixpkgs/pull/335504 was a mistake :laughing:, I think https://github.com/NixOS/nixpkgs/pull/336644 is a much more sensible solution.

It's a minimal change, that should not break anything even for people using `npmInstallHook` manually, but allows us to share code between `npmInstallHook` and `yarnInstallHook`.

**Comment by lelgenio** [comment]
I rebased on master to be able to use the changes from https://github.com/NixOS/nixpkgs/pull/336644.

I'm still working on translating `npm pack` to `yarn` though :thinking:

**Comment by lelgenio** [comment]
Needless to say I'm not happy with the current way `yarn pack` is used.

But also, using `npm pack` for a yarn project has the potential to cause similar problems to what happens now with `npm prune`, unwanted attempts to use the network.

But with `pack` this problem could be even worse since there's no way to opt-out of using it, we need some way to copy the project files.

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: Yea indeed this is a bit weird.. Isn't that doing what `cp -r node_modules` does below?

Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: No, this is installing only the package's files, `cp -r node_modules` is installing only the package's dependencies.

Here's an example of the contents of the dist file for `blade-formatter`, note the absence of `node_modules`:

<details><summary>Details</summary>
<pre>
<code>
$ tar tf blade-formatter.dist.tgz
package
package/CHANGELOG.md
package/LICENSE
package/README.md
package/bin
package/dist
package/package.json
package/src
package/syntaxes
package/bin/blade-formatter.cjs
package/bin/blade-formatter.js
package/dist/types
package/src/cli.ts
package/src/comment.ts
package/src/constants.ts
package/src/errors
package/src/formatter.ts
package/src/indent.ts
package/src/main.ts
package/src/regex.ts
package/src/runtimeConfig.ts
package/src/space.ts
package/src/util.ts
package/src/vsctm.ts
package/syntaxes/blade.tmLanguage.json
package/dist/types/comment.d.ts
package/dist/types/constants.d.ts
package/dist/types/errors
package/dist/types/formatter.d.ts
package/dist/types/indent.d.ts
package/dist/types/main.d.ts
package/dist/types/regex.d.ts
package/dist/types/runtimeConfig.d.ts
package/dist/types/space.d.ts
package/dist/types/util.d.ts
package/dist/types/vsctm.d.ts
package/src/errors/formatError.ts
package/dist/types/errors/formatError.d.ts
</code>
</pre>
</details> 

This retains the same functionality of [these lines](https://github.com/NixOS/nixpkgs/blob/5154507a1b18ae480339201b8f55b145fd52a0f0/pkgs/build-support/node/build-npm-package/hooks/npm-install-hook.sh#L10-L15) from npmInstallHook, but translated to `yarn`.

The problem is that `yarn pack` lacks the `--dry-run` flag, in fact the only flag that it supports is `--filename`: https://classic.yarnpkg.com/en/docs/cli/pack

The `--dry-run` flag does exist, but only in yarn v3+: https://v3.yarnpkg.com/cli/pack

It would be nice to tackle is v3 support too some time: https://github.com/NixOS/nixpkgs/pull/337545/files#diff-1fc5bfea9a2e2f459eb9ad2823a7f1048e5c71e444df55cc774ea114ea5dec23

Comment (doronbehar) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: > It would be nice to tackle is v3 support too some time: [#337545 (files)](https://github.com/NixOS/nixpkgs/pull/337545/files#diff-1fc5bfea9a2e2f459eb9ad2823a7f1048e5c71e444df55cc774ea114ea5dec23)

What exactly should I be looking at there?

File: pkgs/by-name/bl/blade-formatter/package.nix
Comment (doronbehar) [inline pkgs/by-name/bl/blade-formatter/package.nix:40]: BTW This file could be removed - it contains many details related to `package*.json` files.

**Review by doronbehar — APPROVED** [review]
Diff LGTM mostly.

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: People are resorting to making their own FOD in order to use yarn V2, since fetchYarnDeps does not support it.

Comment (doronbehar) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: > People are resorting to making their own FOD in order to use yarn V2, since fetchYarnDeps does not support it.

And there's also yarn v3 available?

Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:23]: v1 is what we support in nixpkgs (https://github.com/yarnpkg/yarn),

All other versions are in https://github.com/yarnpkg/berry

v2 is when changes were made to the lockfile format,
v3 is the current stable version,
v4 is the next version, not yet stable, AFAIK.

File: pkgs/by-name/bl/blade-formatter/package.nix
Comment (lelgenio) [reply pkgs/by-name/bl/blade-formatter/package.nix]: :+1: Thanks, replacing it with `nix-update-script`.

**Comment by lelgenio** [comment]
Hold on a second, got a build failure for `postlight-parser` :thinking:

**Comment by lelgenio** [comment]
Fixed two things:

1. `name` from `package.json` can contain special characters like `/`, so the output of `yarn pack` now goes to a fixed path in `$TMPDIR`.
2. `yarn pack` does not work at all with `bundleDependencies`, from what I gather this feature was useful before lockfiles where a thing, we are providing all dependencies according to `yarn.lock`, so we can just ignore it by temporarily removing them from `package.json`

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 328544` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>blade-formatter</li>
    <li>codefresh</li>
    <li>nixpkgs-manual</li>
    <li>postlight-parser</li>
    <li>yarnInstallHook</li>
  </ul>
</details>

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh]: Mentioning https://github.com/NixOS/nixpkgs/issues/254369 here, and marking conversation as resolved.

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [inline pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh:29]: Here's an idea: Use stdout as the `--filename` (somehow), and pipe that directly to `tar -xzf`..

Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh]: `--filename` does not accept the common `-`, but passing `/dev/stdout` does work.

This ends up creating a corrupted file because yarn still prints other logs to stdout, so I needed to add the `--silent` flag.

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 328544` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>blade-formatter</li>
    <li>codefresh</li>
    <li>nixpkgs-manual</li>
    <li>postlight-parser</li>
    <li>yarnInstallHook</li>
  </ul>
</details>

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (doronbehar) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh]: > `--filename` does not accept the common `-`, but passing `/dev/stdout` does work.
> 
> This ends up creating a corrupted file because yarn still prints other logs to stdout, so I needed to add the `--silent` flag.

Hmm I worry that it is not that platform independent... We should be sure that at least on Darwin it works. A nixpkgs-review report from Darwin will be great...

**Comment by doronbehar** [comment]
There's a small merge conflict now BTW.

**Comment by doronbehar** [comment]
I asked for help on Discourse.

File: pkgs/build-support/node/fetch-yarn-deps/default.nix
Comment (SuperSandro2000) [inline pkgs/build-support/node/fetch-yarn-deps/default.nix:192]: We can use lib.getExe here

**Comment by lelgenio** [comment]
Result of `nixpkgs-review pr 328544` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>blade-formatter</li>
    <li>codefresh</li>
    <li>element-call</li>
    <li>nixpkgs-manual</li>
    <li>postlight-parser</li>
    <li>yarnInstallHook</li>
  </ul>
</details>

File: pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh
Comment (viraptor) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh]: `element-call` fails in the same way.

Comment (lelgenio) [reply pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh]: Reverted to creating a file for now. I'm not sure there are any alternatives.

**Comment by doronbehar** [comment]
@ofborg eval

**Comment by doronbehar** [comment]
@NixOS/nixpkgs-vet something is wrong here I think.

**Review by doronbehar — APPROVED** [review]
Would like to merge when CI is green - waiting for @NixOS/nixpkgs-vet 's response regarding the failure.

**Comment by philiptaron** [comment]
Could you try rebasing on master right now and seeing if the error persists? The tree had some evaluation failures checked in earlier today.

**Comment by doronbehar** [comment]
So I understand there is no equivalent to `@ofborg eval` for the nixpkgs-vet check?

**Comment by philiptaron** [comment]
> So I understand there is no equivalent to `@ofborg eval` for the nixpkgs-vet check?

Since it's "inside" GitHub actions instead of "outside" GitHub Actions, no, it gets fed a revision and it tests that revision only. It's not an active service or a bot in the same way that @ofborg is. It's just a tool that runs and reports.

**Comment by philiptaron** [comment]
I validated that this PR, when rebased on master, passes `nixpkgs-vet`.

**Comment by infinisil** [comment]
It might be possible by clicking the "Re-run jobs" button in the workflow page. Can only be done by committers, and I'm only 90% convinced that it actually picks up fixed base branches. Please give it a try when you see it next time @philiptaron (too late for this PR though because the merge branch is already gone) :)

**Pull request merged**
